types/wlr_cursor.c

```diff
@@ -193,53 +198,89 @@ static void cursor_warp_unchecked(struct wlr_cursor *cur,
 }
 
 /**
- * Get the most specific mapping box for the device in this order:
+ * Gets the intersection of all mapping boxes for this
+ * input device.
  *
- * 1. device geometry mapping
- * 2. device output mapping
- * 3. cursor geometry mapping
- * 4. cursor output mapping
+ * If no box is found, false is returned, otherwise true.
+ * The box is written to the out parameter.
  *
  * Absolute movement for touch and pen devices will be relative to this box and
- * pointer movement will be constrained to this box.
- *
- * If none of these are set, returns NULL and absolute movement should be
+ * pointer movement will be constrained to this box, or if there is no box,
+ * relative to the extents of the layout.
  * relative to the extents of the layout.
  */
-static struct wlr_box *get_mapping(struct wlr_cursor *cur,
-		struct wlr_input_device *dev) {
+static bool get_mapping(struct wlr_cursor *cur,
+		struct wlr_input_device *dev, struct wlr_box *out) {
 	assert(cur->state->layout);
+
+	bool valid = false;
+
 	struct wlr_cursor_device *c_device = get_cursor_device(cur, dev);
 
 	if (c_device) {
 		if (c_device->mapped_box) {
-			return c_device->mapped_box;
+			*out = *c_device->mapped_box;
+			valid = true;
 		}
 		if (c_device->mapped_output) {
-			return wlr_output_layout_get_box(cur->state->layout,
+			struct wlr_box *layout_box = wlr_output_layout_get_box(cur->state->layout,
 				c_device->mapped_output);
+			if (!valid) {
+				*out = *layout_box;
+				valid = true;
+			} else {
+				wlr_box_intersection(out, layout_box, out);
+			}
 		}
 	}
 
 	if (cur->state->mapped_box) {
-		return cur->state->mapped_box;
+		if (!valid) {
+			*out = *cur->state->mapped_box;
+			valid = true;
+		} else {
+			wlr_box_intersection(out, cur->state->mapped_box, out);
+		}
 	}
 	if (cur->state->mapped_output) {
-		return wlr_output_layout_get_box(cur->state->layout,
+		struct wlr_box *layout_box = wlr_output_layout_get_box(cur->state->layout,
 			cur->state->mapped_output);
+		if (!valid) {
+			*out = *layout_box;
+			valid = true;
+		} else {
+			wlr_box_intersection(out, layout_box, out);
+		}
 	}
 
-	return NULL;
+	return valid && !wlr_box_empty(out);
+}
+
+/**
+ * Gets the intersection of all confines for this
+ * input device.
+ * Confines are a superset of mappings.
+ *
+ * This function wraps get_mapping.
+ */
+static bool get_confine(struct wlr_cursor *cur,
+		struct wlr_input_device *dev, struct wlr_box *out) {
+	if (get_mapping(cur, dev, out)) {
+		return wlr_box_intersection(out, &cur->state->confined_box, out);
```

Comment:
It can't be empty.

Reply:
I think I misunderstood your comment as it being null, if it's empty then it won't be correct seemingly, but then it wouldn't work, and it *does* work, so honestly I have no idea, but it doesn't matter now since I'm making it a pointer now.